language/vm/vm-runtime/src/libra_vm.rs

```diff
@@ -1,100 +1,550 @@
-                    result.push(vm.execute_runtime(|runtime| {
+                    result.push(self.move_vm.execute_runtime(|runtime| {
                         process_block_metadata(block_metadata, runtime, state_view, &mut data_cache)
                     })?)
                 }
                 // TODO: Implement the logic for processing writeset transactions.
                 TransactionBlock::WriteSet(_) => unimplemented!(""),
             }
         }
+        report_block_count(count);
+        Ok(result)
+    }
+
+    fn execute_user_transactions(
+        &mut self,
+        txn_block: Vec<SignedTransaction>,
+        data_cache: &mut BlockDataCache<'_>,
+        state_view: &dyn StateView,
+    ) -> VMResult<Vec<TransactionOutput>> {
+        let signature_verified_block: Vec<Result<SignatureCheckedTransaction, VMStatus>> =
+            txn_block
+                .into_par_iter()
+                .map(|txn| {
+                    txn.check_signature()
+                        .map_err(|_| VMStatus::new(StatusCode::INVALID_SIGNATURE))
+                })
+                .collect();
+        let mut result = vec![];
+        for transaction in signature_verified_block {
+            record_stats! {time_hist | TXN_TOTAL_TIME_TAKEN | {
+                    let output = match transaction {
+                        Ok(txn) => self.execute_user_transaction(state_view, data_cache, &txn),
+                        Err(e) => discard_error_output(e),
+                    };
+                    report_execution_status(output.status());
 
+                    // `result` is initially empty, a single element is pushed per loop iteration and
+                    // the number of iterations is bound to the max size of `signature_verified_block`
+                    assume!(result.len() < usize::max_value());
+                    result.push(output);
+                }
+            }
+        }
         Ok(result)
     }
 }
 
+pub(crate) fn discard_error_output(err: VMStatus) -> TransactionOutput {
+    // Since this transaction will be discarded, no writeset will be included.
+    TransactionOutput::new(
+        WriteSet::default(),
+        vec![],
+        0,
+        TransactionStatus::Discard(err),
+    )
+}
+
+// Validators external API
+impl VMVerifier for LibraVM {
+    /// Determine if a transaction is valid. Will return `None` if the transaction is accepted,
+    /// `Some(Err)` if the VM rejects it, with `Err` as an error code. We verify the following
+    /// items:
+    /// 1. The signature on the `SignedTransaction` matches the public key included in the
+    ///    transaction
+    /// 2. The script to be executed is in the whitelist.
+    /// 3. Invokes `LibraAccount.prologue`, which checks properties such as the transaction has the
+    /// right sequence number and the sender has enough balance to pay for the gas.
+    /// 4. Transaction arguments matches the main function's type signature.
```

Comment:
I am not sure we should do this but I guess it's ok if we do

Reply:
The main reason why I'm a bit hesitant to do this is because verifying types will require deserializing bytes into CompiledScript, which is something I'm not sure if needs to be done at ACstep.

Reply:
The main reason why I'm a bit hesitant to do this is because verifying types will require deserializing bytes into CompiledScript, which is something I'm not sure if needs to be done at AC step.